atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraClientPool.java

```diff
@@ -0,0 +1,428 @@
+import java.util.Collection;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.ScheduledThreadPoolExecutor;
+import java.util.concurrent.ThreadLocalRandom;
+import java.util.concurrent.TimeUnit;
+
+import org.apache.cassandra.thrift.Cassandra;
+import org.apache.cassandra.thrift.Cassandra.Client;
+import org.apache.cassandra.thrift.CfDef;
+import org.apache.cassandra.thrift.InvalidRequestException;
+import org.apache.cassandra.thrift.KsDef;
+import org.apache.cassandra.thrift.NotFoundException;
+import org.apache.cassandra.thrift.SchemaDisagreementException;
+import org.apache.cassandra.thrift.TimedOutException;
+import org.apache.cassandra.thrift.TokenRange;
+import org.apache.cassandra.thrift.UnavailableException;
+import org.apache.thrift.TException;
+import org.apache.thrift.transport.TTransportException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.common.base.Function;
+import com.google.common.collect.HashMultimap;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableRangeMap;
+import com.google.common.collect.ImmutableSet;
+import com.google.common.collect.Lists;
+import com.google.common.collect.Maps;
+import com.google.common.collect.Multimap;
+import com.google.common.collect.Range;
+import com.google.common.collect.RangeMap;
+import com.google.common.collect.Sets;
+import com.google.common.collect.Sets.SetView;
+import com.google.common.io.BaseEncoding;
+import com.google.common.primitives.UnsignedBytes;
+import com.google.common.util.concurrent.ThreadFactoryBuilder;
+import com.palantir.atlasdb.cassandra.CassandraKeyValueServiceConfig;
+import com.palantir.atlasdb.keyvalue.api.InsufficientConsistencyException;
+import com.palantir.atlasdb.keyvalue.api.TableReference;
+import com.palantir.common.base.FunctionCheckedException;
+import com.palantir.common.concurrent.PTExecutors;
+
+/**
+ * Feature breakdown:
+ *   - Pooling
+ *   - Token Aware Mapping / Query Routing / Data partitioning
+ *   - Retriable Queries
+ *   - Pool member error tracking / blacklisting*
+ *   - Pool refreshing
+ *   - Pool node autodiscovery
+ *   - Pool member health checking*
+ *
+ *   *entirely new features
+ *
+ *   By our old system, this would be a RefreshingRetriableTokenAwareHealthCheckingManyHostCassandraClientPoolingContainerManager;
+ *   ... this is one of the reasons why there is a new system.
+ **/
+public class CassandraClientPool {
+    private static final Logger log = LoggerFactory.getLogger(CassandraClientPool.class);
+    private static long BANNED_HOST_BACKOFF_TIME_MS = TimeUnit.MINUTES.toMillis(5);
+    private static int MAX_TRIES = 3;
+
+    RangeMap<LightweightOPPToken, List<InetSocketAddress>> tokenMap = ImmutableRangeMap.of();
+    Map<InetSocketAddress, Long> blacklistedHosts = Maps.newHashMap();
+    Map<InetSocketAddress, CassandraClientPoolingContainer> currentPools = Maps.newHashMap();
```

Comment:
We still need to adjust the readers/writers for `currentPools` as the refresh thread is populating it without any visibility guarantees in `getRandomGoodHost()`. Making `currentPools` a CHM likely works, but also worth considering a volatile ref to an ImmutableMap, and do a single volatile read in `getRandomGoodHost()` to a local var.